src/ripple/app/tx/impl/SetOracle.cpp

```diff
@@ -0,0 +1,259 @@
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/SetOracle.h>
+#include <ripple/ledger/Sandbox.h>
+#include <ripple/ledger/View.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Rules.h>
+#include <ripple/protocol/TxFlags.h>
+#include <ripple/protocol/digest.h>
+
+namespace ripple {
+
+static inline uint256
+tokenPairHash(STObject const& pair)
+{
+    return sha512Half(
+        pair.getFieldCurrency(sfBaseAsset).currency(),
+        pair.getFieldCurrency(sfQuoteAsset).currency());
+}
+
+NotTEC
+SetOracle::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featurePriceOracle))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    auto const& dataSeries = ctx.tx.getFieldArray(sfPriceDataSeries);
+    if (dataSeries.size() == 0 || dataSeries.size() > maxOracleDataSeries)
+        return temBAD_ARRAY_SIZE;
+
+    auto invalidLength = [&](auto const& sField, std::size_t length) {
+        return ctx.tx.isFieldPresent(sField) &&
+            ctx.tx[sField].length() > length;
+    };
+
+    if (invalidLength(sfProvider, maxOracleProvider) ||
+        invalidLength(sfURI, maxOracleURI) ||
+        invalidLength(sfAssetClass, maxOracleSymbolClass))
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+SetOracle::preclaim(PreclaimContext const& ctx)
+{
+    auto const sleSetter =
+        ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount)));
+    if (!sleSetter)
+        return terNO_ACCOUNT;
+
+    // lastUpdateTime must be within maxLastUpdateTimeDelta seconds
+    // of the last closed ledger
+    using namespace std::chrono;
+    std::size_t const closeTime =
+        duration_cast<seconds>(ctx.view.info().closeTime.time_since_epoch())
+            .count();
+    std::size_t const lastUpdateTime = ctx.tx[sfLastUpdateTime];
+    if (lastUpdateTime <= epoch_offset.count())
+        return tecINVALID_UPDATE_TIME;
+    std::size_t const lastUpdateTimeEpoch =
+        lastUpdateTime - epoch_offset.count();
+    if (lastUpdateTimeEpoch < closeTime ||
+        lastUpdateTimeEpoch > (closeTime + maxLastUpdateTimeDelta))
+        return tecINVALID_UPDATE_TIME;
+
+    hash_set<uint256> pairs;
+    for (auto const& entry : ctx.tx.getFieldArray(sfPriceDataSeries))
+    {
+        if (!entry.isFieldPresent(sfAssetPrice))
+            return temMALFORMED;
```

Comment:
I wanted to reduce the number of redundant iterations. The other `tem` checks require the ledger access, which is usually done in `preclaim`.

Reply:
I can move some checks to `preflight`: check if `AssetPrice` is included and if there are duplicate entries.
I wouldn't need to repeat this in `preclaim` in case of create but in case of update I still need to iterate over submitted array to see if there are token pairs that are going to be added so I could check the array size and the reserve. The double iteration can't be avoided in case of update. In reality it's not a big deal, since the submitted array size is less or equal to ten. Would you rather have it handled this way? I'd have to iterate in `doApply` but again it's such a small array size that I don't think it matters. As for `tec` and `tem` code, I don't think it's a hard core rule that `tem` can't appear in `preclaim`. For instance, `CreateOffer`, `CancelOffer`, and `NFTokenAcceptOffer` return `tem` in `preclaim`.